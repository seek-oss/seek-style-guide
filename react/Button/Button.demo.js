import React from 'react';
import { Button, HeartIcon, StarIcon } from 'seek-style-guide/react';
import * as sketch from './Button.sketch';
import styles from './Button.less';
import classnames from 'classnames';

export default {
  route: '/button',
  title: 'Button',
  category: 'Form',
  component: Button,
  sketch,
  initialProps: {
    children: 'Hello world',
    color: 'pink'
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Hover',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootHover)
          })
        },
        {
          label: 'Active',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootActive)
          })
        },
        {
          label: 'Focus',
          transformProps: ({ className, ...props }) => ({
            ...props,
            className: classnames(className, styles.rootFocus)
          })
        },
        {
          label: 'Loading',
          transformProps: props => ({
            ...props,
            loading: true
          })
        },
        {
          label: 'Full width',
          transformProps: props => ({
            ...props,
            fullWidth: true
          })
        },
        {
          label: 'Ghost',
          transformProps: props => ({
            ...props,
            ghost: true
          })
        }
      ]
    },
    {
      label: 'Component',
      type: 'radio',
      states: [
        {
          label: 'As Button',
          transformProps: props => ({
            ...props,
            component: 'button'
          })
        },
        {
          label: 'As Link',
          transformProps: props => ({
            ...props,
            component: 'a',
            href: '#'
          })
        },
        {
          label: 'As Div',
          transformProps: props => ({
            ...props,
            component: 'div'
          })
        }
      ]
    },
    {
      label: 'Colour',
      type: 'radio',
      states: [
        {
          label: 'Pink',
          transformProps: props => ({
            ...props,
            color: 'pink'
          })
        },
        {
          label: 'White',
          transformProps: props => ({
            ...props,
            color: 'white'
          })
        },
        {
          label: 'Blue',
          transformProps: props => ({
            ...props,
            color: 'blue'
          })
        },
        {
          label: 'Gray',
          transformProps: props => ({
            ...props,
            color: 'gray'
          })
        },
        {
          label: 'Transparent',
          transformProps: props => ({
            ...props,
            color: 'transparent'
          })
        }
      ]
    },
    {
      label: 'Icon',
      type: 'radio',
      states: [
        {
          label: 'No Icon',
          transformProps: props => props
        },
        {
          label: 'Heart Icon',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <HeartIcon filled={true} />
                &nbsp;Click here
              </span>
            )
          })
        },
        {
          label: 'Star Icon',
          transformProps: props => ({
            ...props,
            children: (
              <span>
                <StarIcon filled={true} />
                &nbsp;Click here
              </span>
            )
          })
        }
      ]
    }
  ]
};
