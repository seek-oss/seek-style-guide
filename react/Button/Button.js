// @flow
import styles from './Button.less';
import React from 'react';
import { Component } from 'react';
import classnames from 'classnames';
import capitalize from 'lodash/capitalize';

type Props = {
  color: 'pink' | 'blue' | 'gray' | 'transparent' | 'white',
  children: React$Node,
  className?: string,
  component?: string | Function,
  ghost?: boolean,
  loading?: boolean,
  fullWidth?: boolean
};

export default class Button extends Component<Props> {
  static displayName = 'Button';

  static defaultProps = {
    className: '',
    ghost: false,
    loading: false,
    fullWidth: false,
    component: 'button'
  };

  button: ?HTMLElement;
  props: Props;

  render() {
    const {
      color,
      ghost,
      className,
      loading,
      fullWidth,
      children,
      component = 'button',
      ...restProps
    } = this.props;

    const combinedProps = {
      className: classnames(styles.root, className, {
        [styles.loading]: loading,
        [styles.fullWidth]: fullWidth,
        [styles.root_isGhost]: ghost && color !== 'transparent',
        [styles[`root_is${capitalize(color)}`]]: color
      }),
      disabled: loading,
      ...restProps
    };

    return React.createElement(component, combinedProps, children);
  }
}
