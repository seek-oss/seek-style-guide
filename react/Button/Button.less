@import (reference) "~seek-style-guide/theme";

@keyframes loading {
  to {
    width: @gutter-width;
  }
}

.hoverState(@rules) {
  &:hover,
  &.rootHover {
    @rules();
  }
}

.activeState(@rules) {
  &:active,
  &.rootActive {
    @rules();
  }
}

.focusState(@rules) {
  &:focus,
  &.rootFocus {
    @rules();
  }
}

.buttonColor(@color, @background-color) {
  color: @color;
  background-color: @background-color;
  .hoverState({background-color: lighten(@background-color, 5%) ;});
  .activeState({background-color: darken(@background-color, 5%) ;});
}

.ghostColor(@color) {
  .activeBackground() {
    background-color: lighten(@color, 51%);
  }

  border-color: @color;
  color: @color;

  .hoverState({background-color: lighten(@color, 56%) ;});

  .activeState({.activeBackground;});

  .focusState({.activeBackground;});
}

.root {
  .touchableText();
  font-weight: @sk-medium;
  display: inline-block;
  text-decoration: none;
  user-select: none;
  text-align: center;
  padding: 0 @gutter-width;
  box-shadow: inset 0 -1px fade(@sk-charcoal, 40%);
  border: 0;
  border-radius: 2px;
  transition: background-color 0.2s, @spring-transition;
  cursor: pointer;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  .activeState({box-shadow: none; outline: none; transform: scale(0.95) ;});
  .focusState({.focus() ;});
}

.root_isPink {
  .buttonColor(@sk-white, @sk-pink);
}

.root_isBlue {
  .buttonColor(@sk-white, @sk-blue-lighter);
}

.root_isGray {
  .buttonColor(@sk-black, @sk-mid-gray-light);
}

.root_isWhite {
  .buttonColor(@sk-black, @sk-white);
}

.root_isGhost,
.root_isTransparent {
  background-color: transparent;
  box-shadow: none;
}

.root_isGhost {
  @border-width: 2px;

  border: @border-width solid;
  line-height: @touchableTextHeight - (@border-width * 2);
}

.root_isPink.root_isGhost {
  .ghostColor(@sk-pink);
}

.root_isBlue.root_isGhost {
  .ghostColor(@sk-blue-lighter);
}

.root_isWhite.root_isGhost {
  border-color: @sk-white;
  color: @sk-white;

  .activeBackground() {
    background-color: rgba(255, 255, 255, 0.2);
  }

  .activeState({.activeBackground;});

  .hoverState({.activeBackground;});

  .focusState({.activeBackground;});
}

.root_isTransparent {
  color: @sk-link;
  padding-left: 0;
  padding-right: 0;

  .hoverState({text-decoration: underline;});

  .activeState({transform: none;});

  .touchableText(@standard-type-scale);
}

.loading {
  padding-right: @gutter-width * 2;

  &::after {
    content: " \2026";
    text-align: left;
    width: 0;
    overflow-x: hidden;
    display: inline-block;
    position: absolute;
    animation-duration: 0.8s;
    animation-name: loading;
    animation-iteration-count: infinite;
  }
}

.fullWidth {
  width: 100%;
}

.group {
  width: 100%;

  .root {
    vertical-align: top;
    @media only screen and (max-width: 500px) {
      width: 100%;
    }
    &:not(.root_isTransparent) {
      @media @mobile {
        padding-left: @gutter-width * 2;
        padding-right: @gutter-width * 2;
      }
    }
    & + .root {
      margin-left: @gutter-width / 2;
      @media only screen and (max-width: 500px) {
        margin-top: (@row-height * 2);
        margin-left: 0;
      }
    }
    & + .root_isTransparent {
      margin-left: @gutter-width * 2;
      @media only screen and (max-width: 500px) {
        margin-left: 0;
      }
    }
  }
}
