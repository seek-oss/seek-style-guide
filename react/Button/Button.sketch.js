import React from 'react';
import Button from './Button';

export const symbols = {
  'Button/Pink': <Button color="pink">Button</Button>,
  'Button/Pink Ghost': (
    <Button color="pink" ghost>
      Button
    </Button>
  ),
  'Button/Blue': <Button color="blue">Button</Button>,
  'Button/Blue Ghost': (
    <Button color="blue" ghost>
      Button
    </Button>
  ),
  'Button/Gray': <Button color="gray">Button</Button>,
  'Button/Gray Ghost': (
    <Button color="gray" ghost>
      Button
    </Button>
  ),
  'Button/Transparent': <Button color="transparent">Button</Button>,
  'Button/White Ghost': (
    <Button color="white" ghost>
      Button
    </Button>
  )
};
