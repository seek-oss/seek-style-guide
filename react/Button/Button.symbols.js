import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

export default class ButtonSymbols extends Component {
  propTypes = {
    SketchSymbol: PropTypes.func.isRequired
  };

  render() {
    const { SketchSymbol } = this.props;

    return (
      <div>
        <SketchSymbol name="Button/pink">
          <Button color="pink">Button</Button>
        </SketchSymbol>

        <SketchSymbol name="Button/blue">
          <Button color="blue">Button</Button>
        </SketchSymbol>

        <SketchSymbol name="Button/transparent">
          <Button color="transparent">Button</Button>
        </SketchSymbol>
      </div>
    );
  }
}
