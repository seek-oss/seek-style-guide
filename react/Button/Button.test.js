import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button:', () => {
  describe('color variants:', () => {
    it('should render blue button', () => {
      const wrapper = shallow(<Button color="blue">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render pink button', () => {
      const wrapper = shallow(<Button color="pink">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render gray button', () => {
      const wrapper = shallow(<Button color="gray">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render white button', () => {
      const wrapper = shallow(<Button color="white">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render transparent button', () => {
      const wrapper = shallow(<Button color="transparent">SEEK</Button>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render white ghost button', () => {
      const wrapper = shallow(
        <Button color="white" ghost>
          SEEK
        </Button>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should render with node', () => {
    const wrapper = shallow(
      <Button color="blue">
        <h5>SEEK</h5>
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with array of nodes', () => {
    const wrapper = shallow(
      <Button color="blue">
        <span>SEEK</span>
        <span>AU</span>
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as an anchor with href="https://www.seek.com.au"', () => {
    const wrapper = shallow(
      <Button color="pink" component="a" href="https://www.seek.com.au">
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with className', () => {
    const wrapper = shallow(
      <Button color="gray" className="foo">
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render fullwidth button', () => {
    const wrapper = shallow(
      <Button color="transparent" fullWidth>
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as loading', () => {
    const wrapper = shallow(
      <Button color="blue" loading>
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render custom component based on itâ€™s reference', () => {
    const CustomComponent = props => <a {...props} someProp="kraken" />;
    const wrapper = shallow(
      <Button color="pink" component={CustomComponent}>
        SEEK
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
