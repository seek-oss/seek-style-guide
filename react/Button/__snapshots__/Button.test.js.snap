// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: color variants: should render blue button 1`] = `
<button
  className="Button__root Button__root_isBlue"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render gray button 1`] = `
<button
  className="Button__root Button__root_isGray"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render pink button 1`] = `
<button
  className="Button__root Button__root_isPink"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render transparent button 1`] = `
<button
  className="Button__root Button__root_isTransparent"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render white button 1`] = `
<button
  className="Button__root Button__root_isWhite"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: color variants: should render white ghost button 1`] = `
<button
  className="Button__root Button__root_isGhost Button__root_isWhite"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: should render as an anchor with href="https://www.seek.com.au" 1`] = `
<a
  className="Button__root Button__root_isPink"
  disabled={false}
  href="https://www.seek.com.au"
>
  SEEK
</a>
`;

exports[`Button: should render as loading 1`] = `
<button
  className="Button__root Button__loading Button__root_isBlue"
  disabled={true}
>
  SEEK
</button>
`;

exports[`Button: should render custom component based on itâ€™s reference 1`] = `
<CustomComponent
  className="Button__root Button__root_isPink"
  disabled={false}
>
  SEEK
</CustomComponent>
`;

exports[`Button: should render fullwidth button 1`] = `
<button
  className="Button__root Button__fullWidth Button__root_isTransparent"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: should render with array of nodes 1`] = `
<button
  className="Button__root Button__root_isBlue"
  disabled={false}
>
  <span>
    SEEK
  </span>
  <span>
    AU
  </span>
</button>
`;

exports[`Button: should render with className 1`] = `
<button
  className="Button__root foo Button__root_isGray"
  disabled={false}
>
  SEEK
</button>
`;

exports[`Button: should render with node 1`] = `
<button
  className="Button__root Button__root_isBlue"
  disabled={false}
>
  <h5>
    SEEK
  </h5>
</button>
`;
